descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h

```diff
@@ -0,0 +1,204 @@
+            /** \brief The definition of the vertex (rung, idx) */
+          std::pair<long unsigned int, long unsigned int> vertex;
+        };
+
+        /** \brief Constructor. The dimension of of the space needs to be specified. A space representing
+            the Descartes State Space will be instantiated */
+        DescartesStateSpace<FloatType>(descartes_light::BGLGraph<FloatType> graph,
+                                       std::vector<std::vector<VertexDesc<FloatType>>> ladder_rungs,
+                                       const std::vector<typename descartes_light::EdgeEvaluator<FloatType>::ConstPtr>& edge_eval,
+                                       const double max_dist,
+                                       const double rung_to_rung_dist = 100000)
+          : graph_(graph)
+          , ladder_rungs_(ladder_rungs)
+          , edge_eval_(std::move(edge_eval))
+          , max_dist_(max_dist)
+          , rung_to_rung_dist_(rung_to_rung_dist)
+          , stateBytes_(sizeof(std::pair<long unsigned int, long unsigned int>))
+        {
+            type_ = 14; // Larger than default types
+            setName("Descartes" + getName());
+        }
+
+        ~DescartesStateSpace() override = default;
+
+        /** \brief This state space is dicsrete which causes certain behaviors in OMPL */
+        bool isDiscrete() const override;
+
+        /** \brief Return the number of dimenstions in this space (2) */
+        unsigned int getDimension() const override;
+
+        /** \brief Get the maximum distance possible to record, this is the rung_to_rung_dist_*(number of rungs) */
+        double getMaximumExtent() const override;
+
+        /** \brief Returns the number of rungs in the ladder */
+        double getMeasure() const override;
+
+        /** \brief Ensures that the state is within the ladder graph */
+        void enforceBounds(ompl::base::State *state) const override;
+
+        /** \brief Checks if the state is within the ladder graph */
+        bool satisfiesBounds(const ompl::base::State *state) const override;
+
+        void copyState(ompl::base::State *destination, const ompl::base::State *source) const override;
+
+        unsigned int getSerializationLength() const override;
+
+        void serialize(void *serialization, const ompl::base::State *state) const override;
+
+        void deserialize(ompl::base::State *state, const void *serialization) const override;
+
+        /** \brief Measures the distance between two given states.
+         * If they are on the same rung this will start with a base distance of 1 rung length apart.
+         * If they are more than 1 rung apart it will have a base distance of rung_to_rung_dist_*(rung difference) + (index difference).
+         * If they are 1 rung apart it will evaluate the cost of the edge.
+         * The rung it is going towards will then have the cost of that vertex
+         * added unless it is going to the goal vertex which has a cost of 0. */
```

Comment:
> Is there an implicit requirement here that the states in each rung be "sorted" such that the same index in each rung represents a similar state (in terms of configuration, etc.)? If so, we should explicitly call out this assumption

The assumption is that these states are not sorted in any way.

> Why? Shouldn't the distance between states on the same rung be infinite because it not conceptually possible to visit multiple states on the same rung

I don't think it actually matters what the distance on the same rung is. As long as it's greater than `max_dist_` and triggers an interpolation. I wanted to set it to 1 rung apart because in my way of thinking while it is impossible to connect [1,2] ([rung,index]) to [1,4], it is also impossible to connect [1,2] to [50,2], but when visualizing those three points in the graph, [1,2] is geometrically much closer to [1,4] than [50,2]. However you are right that [1,2] and [50,2] could both exist in a solution whereas [1,2] and [1,4] could not. In the end I believe both will result in the exact same behavior given how I have the interpolation algorithm set up. The distance just has to be less than the value returned from `getMaximumExtent`.

> I don't understand what you're trying to say here. Please clarify

I will more explicitly explain everything in the description. What I meant here is that the total cost for a connection from point A to point B is the cost of the edge plus the cost of point B. Except for the case where point B is the goal state, which should inherently have a cost of 0, so it is not evaluated.